lib/deref/index.js

```diff
@@ -1,65 +1,62 @@
-var __key = require("./../internal/key");
-var __path = require("./../internal/path");
-var __parent = require("./../internal/parent");
 var InvalidDerefInputError = require("./../errors/InvalidDerefInputError");
+var getCachePosition = require("./../get/getCachePosition");
+var CONTAINER_DOES_NOT_EXIST = "e";
 
 module.exports = function deref(boundJSONArg) {
-    var path = boundJSONArg && boundJSONArg[__path];
 
-    // The simple deref when there is a path.
-    if (path) {
-        return this._clone({
-            _path: path
-        });
-    }
-
-    var key = boundJSONArg && boundJSONArg[__key];
-
-    // This is an invalid argument.  Must* be a generated JSON object from
-    // get.  * - Technically this can just be passed in but don't do that.
-    if (path === undefined && key === undefined) {
-        throw new InvalidDerefInputError();
-    }
-
-    // We have to follow the path back up recursively until we hit a path
-    // or a null parent.
-    var reversedKeys = [key];
-    var reversedLength = 0;
-    var current = boundJSONArg[__parent];
-    path = null;
-    while (current !== null && !path) {
-        key = current[__key];
-        path = current[__path];
+    var absolutePath = boundJSONArg && boundJSONArg.$__path;
+    var refPath = boundJSONArg && boundJSONArg.$__refPath;
+    var toReference = boundJSONArg && boundJSONArg.$__toReference;
+    var referenceContainer;
+
+    // We deref and then ensure that the reference container is attached to
+    // the model.
+    if (absolutePath) {
+        var validContainer = CONTAINER_DOES_NOT_EXIST;
+
+        if (toReference) {
+            validContainer = true;
+            referenceContainer = getCachePosition(this, toReference);
+            // If the reference container is still a sentinel value then compare
+            // the reference value with refPath.  If they are the same, then the
+            // model is still valid.
+            if (refPath && referenceContainer && referenceContainer.$type) {
```

Comment:
Discussed this in person, but just for visibility/anyone else following the PR: Need to make sure validContainer = false, unless "cache has a $type = ref at the same toReference path, and that reference array is the same as refPath".